lib/puppet/type/file/content_uri.rb

```diff
@@ -0,0 +1,42 @@
+require 'net/http'
+require 'uri'
+
+module Puppet
+  Puppet::Type.type(:file).newparam(:content_uri) do
+
+    desc <<-'EOT'
+      A unique identifier for file content, which must be a URI which points
+      to a file in a module. This parameter allows the content of a file resource
+      to remain consistent regardless of code updates on the master.
+
+      The normal form of a `puppet:` URI is:
+
+      `puppet:///modules/<MODULE NAME>/files/<FILE PATH>`
+
+      Note that content_uri contains "files" in its path, unlike the source parameter.
+    EOT
+
+    validate do |path|
+      begin
+        uri = URI.parse(URI.escape(path))
+      rescue
+        self.fail Puppet::Error, "Could not understand URI #{path}: #{detail}", detail
+      end
+
+      self.fail "Cannot use opaque URLs '#{source}'" unless uri.hierarchical?
+      unless uri.scheme == "puppet"
+        self.fail "Must use URLs of type puppet as content URI"
+      end
+    end
+
+    def get_from_puppet_source(source_or_content, &block)
+      request = Puppet::Indirector::Request.new(:code_content, :find, resource[:content_uri], nil, :environment => resource.catalog.environment_instance)
```

Comment:
I forget how this works. This will error if used against a master without a code_content end-point?

It'd be nice to have some sort of spec test for it.